### DISCUSSION TITLE
What's the easiest way to replace a BRICS fragment?

### DISCUSSION THREAD

**Discussion by DaniilBoiko** [rdkit/rdkit#7683] [Q&A]
Hi there,

I have a problem, where I have to take a molecule, decompose into BRICS fragments, take one of them (and it can contain multiple attachment points) and replace it with its analog, that has the same dummy atoms as original fragment. However, from first try this seems to be very challenging.

What's the easiest way to do that?

Thanks!

upd: so far went with using FindBRICSBonds to know how fragments are connected and FragmentOnBRICSBonds to split the molecule while preserving atom order same as in bond list. After that adding new fragment with Chem.CombineMols, reconnecting all bonds except fragment of interest, connecting remaining based on dummy atoms, and removing old fragment and dummy atoms.

**Comment by bp-kelley** [comment]
I'm a bit confused what you are trying to do.  BRICS uses reaction definitions to "break" bonds and also to recombine them.

For example:

```
>>> m = Chem.MolFromSmiles("CNc1ccccc1CC")
>>> res = Chem.FragmentOnBRICSBonds(m)
>>> Chem.MolToSmiles(res)
'[16*]c1ccccc1[16*].[5*]NC.[8*]CC'
```

If we look at rdkit.Chem.BRICS.reactionDefs we see how to recombine these

i.e. 
```
('5','16','-'),
('8', '16', '-'),
```

However, there are two 16s in the output, meaning that we can make two molecules according to BRICS by recombining these fragments with BRICSBuild.  I'm guessing this is really not what you want, but I'll ignore that for the discussion.

So my initially  suggestion would be to independently run BRICSDecompose to build up your fragments, remove the fragments you don't want, and then run BRICSBuild.

Stupid example:

```
>>> m = Chem.MolFromSmiles("CNc1ccccc1CC")
>>> m2 = Chem.MolFromSmiles("CNc1ccncc1CC")
>>> res = Chem.FragmentOnBRICSBonds(m)
>>> res2 = Chem.FragmentOnBRICSBonds(m2)
>>> frags = Chem.GetMolFrags(res, asMols=True)
>>> frags2 = Chem.GetMolFrags(res2, asMols=True)
>>> [Chem.MolToSmiles(f) for f in frags]
['[5*]NC', '[16*]c1ccccc1[16*]', '[8*]CC']
>>> [Chem.MolToSmiles(f) for f in frags2]
['[5*]NC', '[16*]c1ccncc1[16*]', '[8*]CC']
>>> frags = list(frags)
>>> frags[1] = frags2[1]
>>> [Chem.MolToSmiles(m) for m in BRICS.BRICSBuild(frags, scrambleReagents=False, maxDepth=1)]
['CNc1ccncc1NC', 'CCc1ccncc1NC', 'CCc1cnccc1NC', 'CCc1ccncc1CC']
```

Note the maxDepth is important here otherwise you will make a lot of molecules you probably don't want:

```
>>> [Chem.MolToSmiles(m) for m in BRICS.BRICSBuild(frags)]
['CCc1cnccc1NC', 'CNc1ccncc1NC', 'CNc1ccncc1-c1cnccc1NC', 'CCc1ccncc1-c1cnccc1NC', 'CCc1ccncc1-c1ccncc1-c1cnccc1NC', 'CNc1ccncc1-c1ccncc1-c1cnccc1NC', 'CCc1cnccc1-c1ccncc1-c1cnccc1NC', 'CNc1cnccc1-c1ccncc1-c1cnccc1NC', 'CNc1cnccc1-c1cnccc1NC', 'CCc1cnccc1-c1cnccc1NC', 'CCc1cnccc1-c1cnccc1-c1cnccc1NC', 'CNc1cnccc1-c1cnccc1-c1cnccc1NC', 'CCc1ccncc1-c1cnccc1-c1cnccc1NC', 'CCc1ccncc1-c1ccncc1NC', 'CCc1cnccc1-c1cnccc1CC', 'CCc1cnccc1-c1cnccc1-c1cnccc1CC', 'CCc1ccncc1-c1ccncc1-c1ccncc1NC', 'CCc1ccncc1-c1ccncc1-c1cnccc1CC', 'CCc1cnccc1-c1ccncc1-c1ccncc1CC', 'CCc1cnccc1-c1cnccc1-c1ccncc1NC', 'CCc1cnccc1-c1ccncc1-c1ccncc1NC', 'CNc1cnccc1-c1ccncc1-c1ccncc1NC', 'CCc1cnccc1-c1ccncc1-c1cnccc1CC', 'CCc1ccncc1-c1cnccc1-c1ccncc1NC', 'CNc1cnccc1-c1ccncc1NC', 'CCc1cnccc1-c1ccncc1NC', 'CCc1cnccc1-c1ccncc1CC', 'CCc1ccncc1CC', 'CCc1ccncc1NC', 'CCc1ccncc1-c1cnccc1CC']
```

This might be closer to what you desire, however I have a feeling you might be better off with another approach.
